# Numpy Avanzado

## Obtener los índices de los elementos que satisfagan una condición
Hemos trabajado con indexación booleana, sin embargo, a veces será útil obtener los índices de elementos que cumplan con alguna condición.

La función `np.where` ubica las posiciones de un arreglo en donde la condición es verdadera. Veamos

In [1]:
import numpy as np

In [2]:
arr_rand = np.array([8, 8, 3, 7, 7, 0, 4, 2, 5, 2])
print("Arreglo: ", arr_rand)

Arreglo:  [8 8 3 7 7 0 4 2 5 2]


In [7]:
# Obtener índices (posiciones) en los que el valor sea > 5
indices_mayores_5 = np.where(arr_rand > 5)
print("Posiciones en donde el valor > 5:",indices_mayores_5)

Posiciones en donde el valor > 5: (array([0, 1, 3, 4]),)


Ya que tenemos los índices, podemos extraer los elementos usando `take`

In [4]:
arr_rand.take(indices_mayores_5)

array([[8, 8, 7, 7]])

O bien, usamos el vector `indices_mayores_5` para indexar el arreglo

In [8]:
arr_rand[indices_mayores_5]

array([8, 8, 7, 7])

Obtengamos las posiciones de los valores máximo y mínimo dell arreglo

In [9]:
print("Posición del elemento más grande (max):", np.argmax(arr_rand))
print("Posición del elemento más chico  (min):", np.argmin(arr_rand))

Posición del elemento más grande (max): 0
Posición del elemento más chico  (min): 5



---


## Importar y exportar datos de archivos CSV
Una manera muy habitual de importar conjuntos de datos (les llamaremos **datasets** a partir de ahora) es usando ~~la función~~ el método `np.genfromtxt`. Con este método podemos importar datos de URL's de internet o archivos en nuestra computadora, gestionar los valores faltantes, especificar cómo están delimitados los datos, entre otros.

Una versión menos moldeable de esta función es `np.loadtxt`, la cual asume que no hay valores faltantes en el dataset.

Descarguemos un archivo csv a un numpy array.

In [10]:
#apagar la notación científica
np.set_printoptions(suppress=True)
ruta_url = 'https://raw.githubusercontent.com/selva86/datasets/master/Auto.csv'
data = np.genfromtxt(ruta_url, delimiter=',', skip_header=1, filling_values=-999, dtype='float')
data[:3]

array([[  18. ,    8. ,  307. ,  130. , 3504. ,   12. ,   70. ,    1. ,
        -999. ],
       [  15. ,    8. ,  350. ,  165. , 3693. ,   11.5,   70. ,    1. ,
        -999. ],
       [  18. ,    8. ,  318. ,  150. , 3436. ,   11. ,   70. ,    1. ,
        -999. ]])

**Nota: Recordemos que todos los valores de un numpy array deben ser del mismo tipo de dato. Si vemos el archivo Auto.csv en el navegador, veremos que la última columna (name) es de tipo string, y por lo tanto, numpy iba a imporar este valor como np.nan por defecto. Para evitar esto, utilizamos el parámetro `fillin_values` y le asignamos el valor -999 para que coloque un -999 en todos los valores en los que hubiera colocado nan**

¿Cómo hacemos cuando queremos que nuestro arreglo contenga números y texto?
En este caso, debemos especificar que el tipo de dato es `object` o `None`

In [11]:
data2 = np.genfromtxt(ruta_url, delimiter=',', skip_header=1, dtype=None, encoding=None)
data2[:3]

array([(18., 8, 307., 130, 3504, 12. , 70, 1, '"chevrolet chevelle malibu"'),
       (15., 8, 350., 165, 3693, 11.5, 70, 1, '"buick skylark 320"'),
       (18., 8, 318., 150, 3436, 11. , 70, 1, '"plymouth satellite"')],
      dtype=[('f0', '<f8'), ('f1', '<i8'), ('f2', '<f8'), ('f3', '<i8'), ('f4', '<i8'), ('f5', '<f8'), ('f6', '<i8'), ('f7', '<i8'), ('f8', '<U38')])

In [21]:
data3 = np.genfromtxt("autos.csv", delimiter=',', skip_header=1, dtype=float, filling_values=-999)
data3[:3]

array([[  18. ,    8. ,  307. ,  130. , 3504. ,   12. ,   70. ,    1. ,
        -999. ],
       [  15. ,    8. ,  350. ,  165. , 3693. ,   11.5,   70. ,    1. ,
        -999. ],
       [  18. ,    8. ,  318. ,  150. , 3436. ,   11. ,   70. ,    1. ,
        -999. ]])

Súper. Ahora podemos exportar este array a un archivo csv.

In [24]:
np.savetxt("out.csv", data3, delimiter=",")

--- 

## Lectura de imágenes

Una imagen es, a fien de cuentas, una matriz de pixeles. Podemos utilizar numpy para leer y transformar los valores de esta matriz

In [25]:
import matplotlib.pyplot as plt 

In [26]:
img = plt.imread('beagle.jpg')

In [27]:
img.shape

(1200, 1106, 3)

In [29]:
img[:2,:2]


array([[[255, 255, 157],
        [255, 255, 157]],

       [[255, 255, 157],
        [255, 255, 157]]], dtype=uint8)

### ¿Qué estamos viendo aquí?

Una imagen es una matriz de pixeles. 

Cuando tenemos una imagen a color, entonces necesitamos saber de qué color será cada pixel. Aquí es en donde entra en juego el estándar RGB (Red, Green, Blue), el cual dicta que por cada pixel se necesita una combinación de 3 números con valores entre 0 y 255. 

Cada uno de estos tres números indica la intensidad de cada color rojo, verde o azul:

![](rgb.webp)

---

Entonces veamos nuevamente cuál es el primer pixel de nuestar imagen `beagle.jpg`

img[0,0]

In [30]:
img[0,0]

array([255, 255, 157], dtype=uint8)

Tenemos los valores:

| Color   | RGB  | Valor |
| ------- | ---- | ----- |
| Rojo    | R    | 255   |
| Verde   | G    | 255   |
| Azul    | B    | 157   |

In [31]:
img2=plt.imread('salchicha.jpg')

In [32]:
img2.shape

(1414, 2121, 3)

In [33]:
img2[0,0]

array([ 51, 114,   7], dtype=uint8)